ath79: add support for ZiKing CPE46B

lsd-cat opened the pull request:
ZiKing CPE46B is a POE outdoor 2.4ghz device with an integrated directional
antenna. It is low cost and mostly available via Aliexpress, references can
be found at:
- https://forum.openwrt.org/t/anddear-ziking-cpe46b-ar9331-ap121/60383
- https://git.lsd.cat/g/openwrt-cpe46b

Specifications:

- Atheros AR9330
- 32MB of RAM
- 8MB of flash (SPI NOR)
- 1 * 2.4ghz integrated antenna
- 2 * 10/100/1000 ethernet ports (1 POE)
- 3 * Green LEDs controlled by the SoC
- 3 * Green LEDs controlled via GPIO
- 1 * Reset Button controlled via GPIO
- 1 * 4 pin serial header on the PCB
- Outdoor packaging

Flash instruction:
You can use sysupgrade image directly in vendor firmware which is based
on OpenWrt/LEDE. In case of issues with the vendor GUI, the vendor
Telnet console is vulnerable to command injection and can be used to gain
a shell directly on the OEM OpenWrt distribution.
Signed-off-by: Giulio Lorenzo <salveenee@mortemale.org>

ogrorg committed:
ath79: add support for ZiKing CPE46B

ZiKing CPE46B is a POE outdoor 2.4ghz device with an integrated directional
antenna. It is low cost and mostly available via Aliexpress, references can
be found at:
- https://forum.openwrt.org/t/anddear-ziking-cpe46b-ar9331-ap121/60383
- https://git.lsd.cat/g/openwrt-cpe46b

Specifications:

- Atheros AR9330
- 32MB of RAM
- 8MB of flash (SPI NOR)
- 1 * 2.4ghz integrated antenna
- 2 * 10/100/1000 ethernet ports (1 POE)
- 3 * Green LEDs controlled by the SoC
- 3 * Green LEDs controlled via GPIO
- 1 * Reset Button controlled via GPIO
- 1 * 4 pin serial header on the PCB
- Outdoor packaging

Flash instruction:
You can use sysupgrade image directly in vendor firmware which is based
on OpenWrt/LEDE. In case of issues with the vendor GUI, the vendor
Telnet console is vulnerable to command injection and can be used to gain
a shell directly on the OEM OpenWrt distribution.
Signed-off-by: Giulio Lorenzo <salveenee@mortemale.org>

adschm commented on target/linux/ath79/dts/ar9330_ziking_cpe46b.dts:
```diff
@@ -0,0 +1,117 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
```
dts-v1 can be dropped

adschm commented on target/linux/ath79/dts/ar9330_ziking_cpe46b.dts:
```diff
@@ -0,0 +1,117 @@
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+#include "ar9330.dtsi"
```
This include should be before the others (check recently changed DTS files).

adschm commented on target/linux/ath79/dts/ar9330_ziking_cpe46b.dts:
```diff
@@ -0,0 +1,117 @@
+                compatible = "gpio-leds";
+
+                wlan {
+                        label = "ziking-cpe46b:green:wlan";
```
model prefix can be dropped, keeping "green:wlan" etc.

adschm commented on target/linux/ath79/dts/ar9330_ziking_cpe46b.dts:
```diff
@@ -0,0 +1,117 @@
+
+&gpio {
+        status = "okay";
+};
```
This block can be dropped, gpio is not disabled anywhere.

adschm commented on target/linux/ath79/dts/ar9330_ziking_cpe46b.dts:
```diff
@@ -0,0 +1,117 @@
+        status = "okay";
+};
+
+
```
drop second empty line.

adschm commented on target/linux/ath79/dts/ar9330_ziking_cpe46b.dts:
```diff
@@ -0,0 +1,117 @@
+
+                        partition@1 {
+                                label = "u-boot-env";
+                                reg = <0x010000 0x010000>;
```
You don't define anything in uboot-envtools, so this should be readonly.

lsd-cat replied:
Thank you, got this one too

adschm commented on target/linux/ath79/dts/ar9330_ziking_cpe46b.dts:
```diff
@@ -0,0 +1,117 @@
+                                read-only;
+                        };
+
+                        partition@1 {
```
Use offset here and below, e.g. partition@10000

adschm commented on target/linux/ath79/dts/ar9330_ziking_cpe46b.dts:
```diff
@@ -0,0 +1,117 @@
+&eth1 {
+        status = "okay";
+
+        mtd-mac-address = <&art 0x6>;
```
does this correspond to vendor MAC assignment?

lsd-cat replied:
I have actually an issue there, the vendor shipped ART partition has a placeholder mac address there ("0a  0b 0c 0d 0e 01"). If I run "progmac" from u-boot then it gets rewritten with a valid one (although not the one in the label). I think original vendor firmware just did wlan_mac+1 (since eth0 and wlan mac are as an example :00 and :01 I expect eth2 would be :02). Can you point me in the right direction to fix this?

adschm commented on target/linux/ath79/dts/ar9330_ziking_cpe46b.dts:
```diff
@@ -0,0 +1,117 @@
+        status = "okay";
+
+        mtd-cal-data = <&art 0x1000>;
+        mtd-mac-address = <&art 0x0>;
```
No valid address in 0x1002?

lsd-cat replied:
You're right the correct address it at 0x1002, thank you!

adschm commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1683,6 +1683,16 @@ define Device/yuncore_xd4200
+  DEVICE_VENDOR := ZiKing
+  DEVICE_MODEL := CPE46B
+  IMAGE_SIZE := 8000k
+  SUPPORTED_DEVICES += cpe46b
```
Is this for vendor firmware?

lsd-cat replied:
Do you mean the image size? It is actually the net of 8192k-64k(uboot)-64k(u-boot-env)-64k(art)

adschm replied:
No, I mean SUPPORTED_DEVICES

lsd-cat replied:
You're right, I'll drop the line

adschm replied:
How can I be right when I ask a question?

lsd-cat replied:
You're right that that line is there for no other purpose than a leftover while copying from another target. Are the variables there documented anywhere?

adschm replied:
So, with the line dropped, I suspect one will need to force upgrade now coming from the vendor firmware?

lsd-cat replied:
I'm getting a secondary device with stock firmware in order to test installing openwrt directly from the vendor updater

lsd-cat commented:
Thank you for pointing out the issues. I should have fixed them now:
 - Fixed formatting
 - Added correct MAC Address handling
 - Added target for uboot env tools

adschm commented on target/linux/ath79/dts/ar9330_ziking_cpe46b.dts:
```diff
@@ -0,0 +1,111 @@
+&spi {
+        status = "okay";
+
+        num-chipselects = <1>;
```
This should be dropped.

adschm commented on target/linux/ath79/dts/ar9330_ziking_cpe46b.dts:
```diff
@@ -0,0 +1,111 @@
+
+        aliases {
+                serial0 = &uart;
+                label-mac-device = &eth0;
```
Is this actually what's on the label?

lsd-cat replied:
Here is a hi res picture of the label: https://git.lsd.cat/g/openwrt-cpe46b/raw/master/images/label.jpg
Inside the firmware there are multiple references to "ZiKing" and the IANA code: https://oidref.com/1.3.6.1.4.1.37260
There are multiple resellers on Aliexpress and other sites not stating the brand or using their own one. The one in the firmware looked like the actual manufacturer but it's hard to be sure.
Here are a few:

 * https://www.aliexpress.com/item/32974861845.html
 * https://deals.jumia.ci/antenne-wifi-cpe-46b-pid5819548
 * https://m.afzhan.com/chanpin/10318793.html

lsd-cat replied:
The first address in my ART partition matches the label on the device mainboard, so I guess it's the wrong interface and it should be "wifi0", or isn't it?

adschm replied:
Well, the "first address in [your] ART partition" is 0x0?
And that's assigned to &eth0?

lsd-cat replied:
Sorry for the confusion. Yes the address on the label is at 0x0 in the ART partition. In the u-boot log the same address is listed as eth0.

From stock firmware boot log:
```

U-Boot 1.1.413 (Aug 29 2012 - 10:36:47)

AP121-2MB (ar9330) U-boot
DRAM:  32 MB
flash size 8388608, sector count = 128
Flash:  8 MB
In:    serial
Out:   serial
Err:   serial
Net:
eth0: <address at 0x0 in ART>
```

adschm commented on target/linux/ath79/dts/ar9330_ziking_cpe46b.dts:
```diff
@@ -0,0 +1,111 @@
+        status = "okay";
+
+        mtd-cal-data = <&art 0x1000>;
+        mtd-mac-address = <&art 0x1002>;
```
mtd-mac-address can be dropped, it's read automatically from the cal-data ...

lsd-cat replied:
Ok thanks

adschm commented:
Since this is ancient and was quite easy to fix, I fixed the remaining issues myself and will merge this tonight.
